File: Course_1-Algorithmic_Toolbox/Week-6/Excercise_Challenges/1_maximum_amount_of_gold/knapsack.cpp
#include <iostream>
#include <vector>
using namespace std;

int optimal_weight(int W, vector<int> &w)
{
    int n = w.size();
    vector<vector<int> > dp(n + 1, vector<int>(W + 1));

    for(int i = 0; i < n + 1; ++i){
        for(int j = 0; j < W + 1; ++j){
            if(i == 0 || j == 0)
                dp[i][j] = 0;
            else if(w[i - 1] <= j)
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w[i - 1]] + w[i - 1]);
            else
                dp[i][j] = dp[i - 1][j];
        }
    }

    return dp[n][W];
}

int main(void)
{
    int n, W;
    cin >> W >> n;
    vector<int> w(n);
    for(int i = 0; i < n; i++){
        cin >> w[i];
    }
    cout << optimal_weight(W, w) << '\n';

    return 0;
}

File: Course_1-Algorithmic_Toolbox/Week-6/Excercise_Challenges/1_maximum_amount_of_gold/knapsack.py
import numpy as np

def optimal_weight(W, w):
    n = len(w)
    dp = np.zeros((n + 1, W + 1))

    for i in range(1, n + 1):
        for j in range(1, W + 1):
            if w[i - 1] <= j:
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w[i - 1]] + w[i - 1])
            else:
                dp[i][j] = dp[i - 1][j]

    return int(dp[n][W])

if __name__ == '__main__':
    W, n = map(int, input().split())
    w = list(map(int, input().split()))
    print(optimal_weight(W, w))
